fix: Preserve the order of a context

saig0 opened the pull request:
## Description

Fix an issue that did not store the context entries in the same order as defined in the literal (i.e. `{a: 1, b: 2}` is stored as `{b: 2, a: 1}`). The issue could be observed by using the `get entries()` function.

The root cause was not in the context functions but in how the context is created in the engine via `EvalContext`. Fixing the behavior by using a sorted map for storing the context entries and by avoiding the nesting of the `VariableProvider`s.

As a result, the code of the `EvalContext` is a bit more complex but easier to debug. All variables and functions are stored in one provider instead of a chain with one provider per variable/function.

## Related issues

closes #161 
based on a previous try #340

remcowesterhoud commented on src/main/scala/org/camunda/feel/impl/interpreter/EvalContext.scala:
```diff
@@ -35,77 +38,133 @@ class EvalContext(val valueMapper: ValueMapper,
+  }
+
+  private def toSortedVariableProvider(variables: Map[String, Any]): VariableProvider =
+    StaticVariableProvider(SortedMap[String, Any]() ++ variables)
```
❌ As discussed f2f, this [SortedMap](https://www.scala-lang.org/api/2.13.6/scala/collection/immutable/SortedMap.html) does not behave the way you think it does. It doesn't keep the entries sorted in the order they were added. Instead they are sorted by their key, regardless of when they're added to the map.

Let's also add a test case for this scenario 🙂

remcowesterhoud commented on src/main/scala/org/camunda/feel/impl/interpreter/EvalContext.scala:
```diff
@@ -35,77 +38,133 @@ class EvalContext(val valueMapper: ValueMapper,
+    StaticVariableProvider(SortedMap[String, Any]() ++ variables)
+
+  private def toSortedVariableProvider(entry: (String, Any)): VariableProvider =
+    StaticVariableProvider(SortedMap(entry))
```
❌ Same as above

remcowesterhoud reviewed (changes requested):
Thanks for hopping on a call to explain the changes further @saig0! Please have a look at my comments!

saig0 commented:
@remcowesterhoud great catch. :+1: 

I fixed the issue by using a [SeqMap](https://www.scala-lang.org/api/2.13.6/scala/collection/SeqMap.html). :laughing: Additionally, I added a few more test cases with non-alphabetic order. :see_no_evil: 

Please have a look. :eyes:

remcowesterhoud reviewed (approved):
Thanks @saig0! 🚀

remcowesterhoud commented:
bors merge
